api/src/utils/get-schema.ts

```diff
@@ -31,33 +33,44 @@ export async function getSchema(options?: {
 	const database = options?.database || getDatabase();
 	const schemaInspector = createInspector(database);
 
-	let result: SchemaOverview;
+	if (options?.bypassCache || env['CACHE_SCHEMA'] === false) {
+		return await getDatabaseSchema(database, schemaInspector);
+	}
 
-	if (!options?.bypassCache && env['CACHE_SCHEMA'] !== false) {
-		let cachedSchema;
+	const cached = await getSchemaCache();
 
-		try {
-			cachedSchema = await getSchemaCache();
-		} catch (err: any) {
-			logger.warn(err, `[schema-cache] Couldn't retrieve cache. ${err}`);
-		}
+	if (cached) {
+		return cached;
+	}
 
-		if (cachedSchema) {
-			result = cachedSchema;
-		} else {
-			result = await getDatabaseSchema(database, schemaInspector);
+	const lock = useLock();
+	const bus = useBus();
 
-			try {
-				await setSchemaCache(result);
-			} catch (err: any) {
-				logger.warn(err, `[schema-cache] Couldn't save cache. ${err}`);
-			}
-		}
-	} else {
-		result = await getDatabaseSchema(database, schemaInspector);
+	const lockKey = 'schemaCache--preparing';
+	const messageKey = 'schemaCache--done';
+	const processId = await lock.increment(lockKey);
+
+	const currentProcessShouldHandleOperation = processId === 1;
+
+	if (currentProcessShouldHandleOperation === false) {
+		logger.trace('Schema cache is prepared elsewhere, waiting for result.');
+
+		return new Promise((resolve) => {
+			bus.subscribe(messageKey, async () => {
+				const schema = await getSchema(options);
+				resolve(schema);
+			});
+		});
```

Comment:
We should unsubscribe as this callback is for one-time use but it gets recreated repeatedly.

```suggestion
		return new Promise((resolve) => {
			const subscriptionCallback = async () => {
				await bus.unsubscribe(messageKey, subscriptionCallback);

				const schema = await getSchema(options);
				resolve(schema);
			};

			bus.subscribe(messageKey, subscriptionCallback);
		});
```

We could potentially extract this one-time subscription as a util function to reuse for similar cases. Naming is hard. 🤔